test/common.py

```diff
@@ -1763,19 +1764,36 @@ def clear(self):
     if shared.EMSCRIPTEN_TEMP_DIR:
       utils.delete_contents(shared.EMSCRIPTEN_TEMP_DIR)
 
-  def run_process(self, cmd, check=True, **args):
+  def run_process(self, cmd, check=True, **kwargs):
     # Wrapper around shared.run_process.  This is desirable so that the tests
     # can fail (in the unittest sense) rather than error'ing.
     # In the long run it would nice to completely remove the dependency on
     # core emscripten code (shared.py) here.
+
+    # Handle buffering for subprocesses.  The python unittest buffering mechanism
+    # will only buffer output from the current process (by overwriding sys.stdout
+    # and sys.stderr), not from sub-processes.
+    stdout_buffering = 'stdout' not in kwargs and isinstance(sys.stdout, io.StringIO)
+    stderr_buffering = 'stderr' not in kwargs and isinstance(sys.stderr, io.StringIO)
+    if stdout_buffering:
+      kwargs['stdout'] = PIPE
+    if stderr_buffering:
+      kwargs['stderr'] = PIPE
+
     try:
-      return shared.run_process(cmd, check=check, **args)
+      rtn = shared.run_process(cmd, check=check, **kwargs)
     except subprocess.CalledProcessError as e:
       if check and e.returncode != 0:
         print(e.stdout)
         print(e.stderr)
         self.fail(f'subprocess exited with non-zero return code({e.returncode}): `{shlex.join(cmd)}`')
 
+    if stdout_buffering:
+      sys.stdout.write(rtn.stdout)
+    if stderr_buffering:
+      sys.stderr.write(rtn.stderr)
```

Comment:
I don't understand the change here... if we don't do (say) stdout buffering then we let `run_process` write to stdout directly. otoh if we do stdout buffering then we write that buffer to `sys.stdout` ourselves. Isn't the result the same, maybe only with some slight delay?

Reply:
The issue is that the unittest.py module overrides `sys.stdout` and `sys.stderr` in the expectation that will capture all output: https://github.com/python/cpython/blob/aafb1435d803c68ea3bbd2748f1e141c9deaadb0/Lib/unittest/result.py#L65-L71

However, overriding these doesn't effect subprocess, and the default of `stdout=None, stderr=None` means that the actually process's stdout and stderr are inherited not the user-space overrides of `sys.stdout` and `sys.stderr`

Reply:
Yes, exactly, hopefully the PR description capture that somwhat.